File: old/2020/17_conway_cubes/aoc17.jl
# Conway Cubes
#
# Advent of Code 2020, day 17
# Solution by Geir Arne Hjelle, 2021-01-01

module AOC17

using Pipe

function setup_cubes(initial_state, offset; num_dims)
    cubes = falses(fill(2 * offset + size(initial_state, 1), num_dims)...)
    for idx ∈ findall(initial_state)
        row, col = idx.I
        cubes[row + offset, col + offset, fill(offset + 2, num_dims - 2)...] = true
    end
    cubes
end

function evolve!(num_steps)
    function _evolve!(cubes)
        for _ in 1:num_steps
            next_gen = cubes |> copy
            for idx in CartesianIndices(cubes)[((2:s - 1) for s in size(cubes))...]
                num_active = cubes[([-1, 0, 1] .+ i for i ∈ idx.I)...] |> count  # Includes the cell itself
                next_gen[idx] = cubes[idx] ? num_active ∈ [3, 4] : num_active == 3
            end
            cubes = next_gen
        end
        cubes
    end
end

# Solve the problem for one file
function solve(input)
    # Parse input
    initial_state = @pipe (
        input
        |> split
        .|> map(==('#'), collect(_))
        |> hcat(_...)
        |> permutedims
    )
    num_steps = 6

    # Part 1
    part_1 = setup_cubes(initial_state, num_steps, num_dims=3) |> evolve!(num_steps) |> count

    # Part 2
    part_2 = setup_cubes(initial_state, num_steps, num_dims=4) |> evolve!(num_steps) |> count

    part_1, part_2
end


# Solve the problem for one file
function solve_file(file_path)
    println("\n$(file_path)")
    input = open(file_path) do fid
        read(fid, String) |> strip
    end
    input .|> solve
end

# Solve the problem for each file
[a for a in ARGS if a[1] != '-'] .|> solve_file .|> s -> join(s, "\n") |> println

end  # module

File: old/2020/17_conway_cubes/aoc17.py
"""Conway Cubes

Advent of Code 2020, day 17
Solution by Geir Arne Hjelle, 2020-12-17
"""
# Standard library imports
import itertools
import pathlib
import sys

# Third party imports
import numpy as np

debug = print if "--debug" in sys.argv else lambda *_: None


def parse_cells(text):
    return np.array([[0 if c == "." else 1 for c in ln] for ln in text.split("\n")])


def cells_to_nd(center, ndim, max_generations=6):
    cube = np.zeros(
        np.array(center.shape + (1,) * (ndim - 2)) + 2 * (max_generations + 1),
        dtype=np.uint8,
    )
    # Initialize center plane
    cx, cy, cz = np.array(cube.shape)[:3] // 2
    dx1, dy1 = np.array(center.shape) // 2
    dx2, dy2 = (np.array(center.shape) + 1) // 2
    plane = (
        slice(cx - dx1, cx + dx2),
        slice(cy - dy1, cy + dy2),
    ) + ((slice(cz, cz + 1),) * (ndim - 2))
    cube[plane] = center.reshape(cube[plane].shape)

    return cube


def next_gen(current):
    ndim = current.ndim
    center = (slice(1, -1),) * ndim

    # Count neighbors
    neighbors = np.zeros(current.shape, dtype=np.uint8)
    slices = itertools.product(
        *[[slice(None, -2), slice(1, -1), slice(2, None)]] * ndim
    )
    neighbors[center] = sum(current[s] for s in slices) - current[center]

    # - Active cube: 2 or 3 neighbors active ==> Active
    # - Inactive cube: 3 neighbors active ==> Active
    return ((neighbors == 3) | ((neighbors == 2) & (current == 1))).astype(np.uint8)


def main(args):
    """Solve the problem for all file paths"""
    for file_path in [pathlib.Path(p) for p in args if not p.startswith("-")]:
        solve(file_path)


def solve(file_path):
    """Solve the problem for one file path"""
    print(f"\n{file_path}:")
    center = parse_cells(file_path.read_text().strip())

    # Part 1
    cube = cells_to_nd(center, ndim=3, max_generations=6)
    for _ in range(6):
        cube = next_gen(cube)
    print(f"Active cells in cube after 6 generations: {cube.sum()}")

    # Part 2
    hypercube = cells_to_nd(center, ndim=4, max_generations=6)
    for _ in range(6):
        hypercube = next_gen(hypercube)
    print(f"Active cells in hypercube after 6 generations: {hypercube.sum()}")


if __name__ == "__main__":
    main(sys.argv[1:])
